### PR TITLE
Add Normalization.adapt() support for Python iterables

### PR DESCRIPTION
**Pull request by sharpenteeth**
## Summary
Extends `Normalization.adapt()` to accept arbitrary Python iterables of batches (e.g. lists of arrays, generators), in addition to single arrays and datasets.

## Changes
- **adapt()** now detects iterables via a small helper and consumes them in a single pass, inferring input shape from the first batch.
- Added **_is_iterable_of_batches()** and **_extract_batch()** for safe detection and first-element extraction.
- Clear errors for empty iterables, incompatible batch shapes, and items that are not array-like (e.g. missing `.shape`).
- Docstring updated to document the new behavior.

## Testing
- New tests for adapting from lists of batches, generators, single-batch iterables, and high-dim with `axis` tuple.
- Edge-case tests: empty iterable, incompatible shapes across batches, non-array item in iterable.

Net: +200 insertions, -9 deletions (code-only).

### CODE REVIEW

**Comment by dive2tech** [comment]
Hi, @gbaned
Could you review my PR, please?
Thanks.

**Comment by dive2tech** [comment]
Hi, @gbaned
Could you review my PR, please?
Thanks.

**Comment by dive2tech** [comment]
Hi, @hertschuh 
Could you review my PR, please?
Thanks.

**Comment by dive2tech** [comment]
Hi, @hertschuh
Could you review my PR, please?
Thanks.

**Comment by dive2tech** [comment]
Hi, @hertschuh
Could you review my PR, please?
Thanks.

**Comment by sharpenteeth** [comment]
Hi, @hertschuh
Could you review my PR, please?
Thanks.

File: keras/src/layers/preprocessing/normalization.py [hunk 1]
```diff
@@ -309,16 +358,33 @@ def get_input_shape(d):
...
-        elif isinstance(data, (tf.data.Dataset, PyDataset)):
+        elif isinstance(data, (tf.data.Dataset, PyDataset)) or data_is_iterable:
             total_mean = ops.zeros(self._mean_and_var_shape)
             total_var = ops.zeros(self._mean_and_var_shape)
             total_count = 0
             for batch in data:
+                if data_is_iterable:
+                    batch = _extract_batch(batch)
                 if isinstance(batch, tuple):
                     batch = batch[0]
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization.py]: Replace lines 366-369 with `batch = _extract_batch(batch)`. That's what lines 368-369 were already doing, and whether `data_is_iterable` is irrelevant.

File: keras/src/layers/preprocessing/normalization.py
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization.py]: Replace with `_extract_batch(data)`

File: keras/src/layers/preprocessing/normalization.py [hunk 2]
```diff
@@ -10,6 +11,26 @@
...
+    """True if data is an iterable of batches, not Dataset/array/tensor."""
+    if isinstance(data, (np.ndarray, str, bytes)):
+        return False
+    if backend.is_tensor(data):
+        return False
+    if tf is not None and isinstance(data, tf.data.Dataset):
+        return False
+    if isinstance(data, PyDataset):
+        return False
+    return hasattr(data, "__iter__") and not hasattr(data, "shape")
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization.py]: You don't need all of this. You're using this function in an `elif` where the other `if`s already handle all of the types for which you return False.

All you need is this line 310:
```python
elif hasattr(data, "__iter__"):
```

File: keras/src/layers/preprocessing/normalization.py [hunk 3]
```diff
@@ -309,16 +358,33 @@ def get_input_shape(d):
...
             total_count = 0
             for batch in data:
+                if data_is_iterable:
+                    batch = _extract_batch(batch)
                 if isinstance(batch, tuple):
                     batch = batch[0]
                 batch = backend.convert_to_tensor(
                     batch, dtype=self.compute_dtype
                 )
+                if self.built and data_is_iterable:
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization.py]: You don't need to check `self.built`, lines 343-344 already take care of this.

File: keras/src/layers/preprocessing/normalization.py [hunk 4]
```diff
@@ -309,16 +358,33 @@ def get_input_shape(d):
...
+                            batch_dim is not None
+                            and expected is not None
+                            and batch_dim != expected
+                        ):
+                            raise ValueError(
+                                "adapt() iterable yielded a batch with "
+                                "incompatible shape. Expected "
+                                f"{self._build_input_shape}, got "
+                                f"{tuple(batch.shape)}."
+                            )
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization.py]: I don't see any particular reason to put this in a `if data_is_iterable` block. The same check should pass with `tf.data.Dataset` and `PyDataset` and they are applicable in the sense that those can also return tensors of different shapes.

However, do you know what happens if there is a shape mismatch without this code? Don't you already get an error message?

File: keras/src/layers/preprocessing/normalization_test.py [hunk 5]
```diff
@@ -280,6 +280,142 @@ def test_scalar_broadcast(self):
...
+        layer.adapt(batches)
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=(0, 3)), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=(0, 3)), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
+    def test_adapt_iterator_of_batches(self):
+        x = np.random.random((24, 3)).astype("float32")
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Can you also add a test with a PyGrain dataset?

**Review by hertschuh — COMMENTED** [review]
Thanks for the adding this!

**Comment by sharpenteeth** [comment]
Hi, @hertschuh
Thanks for your feedback.
I refactored code.
Could you review my PR, please?
Thank you for your time.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 6]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+        list_of_batches = [x[i : i + 6] for i in range(0, 24, 6)]
+        layer = layers.Normalization(axis=-1)
+        layer.adapt(iter(list_of_batches))
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=0), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=0), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py:411]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 7]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+        batches = [x[i : i + 8] for i in range(0, 32, 8)]
+        layer = layers.Normalization(axis=(1, 2))
+        layer.adapt(batches)
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=(0, 3)), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=(0, 3)), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 8]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+        x = np.random.random((16, 4)).astype("float32")
+        layer = layers.Normalization(axis=-1)
+        layer.adapt([x])
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=0), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=0), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 9]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+    def test_adapt_iterable_batch_without_shape_raises(self):
+        layer = layers.Normalization(axis=-1)
+
+        def gen_no_shape():
+            yield 42
+
+        with self.assertRaisesRegex(TypeError, "`.shape`"):
+            layer.adapt(gen_no_shape())
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 10]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+        batches = [(x[i : i + 8], np.zeros(8)) for i in range(0, 24, 8)]
+        layer = layers.Normalization(axis=-1)
+        layer.adapt(batches)
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=0), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=0), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 11]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+        layer_list.adapt(list_of_batches)
+        layer_ndarray = layers.Normalization(axis=-1)
+        layer_ndarray.adapt(x)
+        out_list = layer_list(x[:10])
+        out_ndarray = layer_ndarray(x[:10])
+        out_list = backend.convert_to_numpy(out_list)
+        out_ndarray = backend.convert_to_numpy(out_ndarray)
+        self.assertAllClose(out_list, out_ndarray, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 12]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+
+        layer = layers.Normalization(axis=-1)
+        layer.adapt(batch_gen())
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=0), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=0), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 13]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
...
+        batches = [x[:8], x[8:16], x[16:24], x[24:32]]
+        layer = layers.Normalization(axis=-1)
+        layer.adapt(batches)
+        self.assertTrue(layer.built)
+        output = layer(x)
+        output = backend.convert_to_numpy(output)
+        self.assertAllClose(np.var(output, axis=0), 1.0, atol=1e-5)
+        self.assertAllClose(np.mean(output, axis=0), 0.0, atol=1e-5)
+
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

File: keras/src/layers/preprocessing/normalization_test.py [hunk 14]
```diff
@@ -280,6 +280,155 @@ def test_scalar_broadcast(self):
         self.assertEqual(tuple(layer.mean.shape), (1, 7))
         self.assertAllClose(layer.mean, [[5.0] * 7])
 
+    @pytest.mark.requires_trainable_backend
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/normalization_test.py]: Remove `@pytest.mark.requires_trainable_backend`, this should work on all backend AFAICT.

**Review by hertschuh — COMMENTED** [review]
Thanks! Last thing about the unit tests:

**Comment by sharpenteeth** [comment]
@hertschuh, I updated code for your feedback.
Could you review my PR, please?
Thanks.

**Comment by sharpenteeth** [comment]
Hi, @hertschuh.
ready to merge.
